singularity-eos/eos/modifiers/eos_unitsystem.hpp

```diff
@@ -0,0 +1,250 @@
+//------------------------------------------------------------------------------
+// © 2021. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_UNITSYSTEM_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_UNITSYSTEM_HPP_
+
+#include <iostream>
+#include "stdio.h"
+#include <cassert>
+#include <cmath>
+#include <cstdlib>
+#include <limits>
+#include <utility>
+
+#include <ports-of-call/portability.hpp>
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+
+namespace singularity {
+
+// tag dispatch for constructors for UnitSystem
+namespace eos_units_init {
+static struct ThermalUnitsInit { } thermal_units_init_tag;
+static struct LengthTimeUnitsInit { } length_time_units_init_tag;
+} // namespace eos_units
+
+template<typename T>
+class UnitSystem {
+public:
+  // move semantics ensures dynamic memory comes along for the ride
+  // TODO(JMM): Entropy unit needed?
+  UnitSystem(T &&t, eos_units_init::ThermalUnitsInit,
+	     const Real rho_unit, const Real sie_unit, const Real temp_unit)
+    : t_(std::forward<T>(t))
+    , rho_unit_(rho_unit)
+    , sie_unit_(sie_unit)
+    , temp_unit_(temp_unit)
+    , press_unit_(rho_unit*sie_unit)
+    , inv_rho_unit_(1/rho_unit) // inverses computed to avoid division at runtime
+    , inv_sie_unit_(1/sie_unit)
+    , inv_temp_unit_(1/temp_unit)
+    , inv_press_unit_(1/press_unit_)
+    , inv_dpde_unit_(sie_unit/press_unit_) // thermo derivatives computed consistently
+    , inv_dvdt_unit_(rho_unit*temp_unit)  // TODO(JMM): Is this convention weird?
+    , inv_dpdr_unit_(rho_unit/press_unit_)
+    , inv_dtdr_unit_(rho_unit/temp_unit)
+    , inv_dtde_unit_(sie_unit/temp_unit) // obviously this is also Cv
+    , inv_cv_unit_(temp_unit/sie_unit)
+    , inv_gruneisen_unit_(rho_unit*sie_unit/press_unit_)
```

Comment:
Since `press_unit_` is defined as `rho_unit * sie_unit`, this will always evaluate to 1. Which makes sense since the Gruneisen parameter is unitless. Up to you whether you feel it necessary to explicitly calculate this ;-)

Reply:
Oh good catch. Initially I allowed `press_unit_` to be inconsistent with `rho_unit` and `sie_unit`. I think I should remove this to save memory. :)